Death Nettle changes

ViceEmargo opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

This is my first PR, let's gooooo.

## About the PR
<!-- What did you change in this PR? -->
A user must have Botanist's Leather Gloves equipped in order to pick up this feisty plant, or you'll face its stinging wrath.
An unprepared user will be stung, taking heat and caustic damage, with a small chance to get stunned from the pain.
Death Nettle is now able to pierce hard suits.
Death Nettle will also "wilt" after 5 uses.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
Death Nettle is a very powerful tool which Botany must cooperate with Chemistry and Cargo in order to obtain.
Being a melee weapon with a relatively high base damage (8 heat, 8 caustic, totaling 16), as well as being a chemical injector capable of injecting 5u of reagent to a target per swing, a single Death Nettle can be a powerful weapon.
Once a botanist has groomed a Death Nettle into an incredible weapon, it could then be mass produced and distributed to the station in response of Nuclear Operatives or other major station threats.

This PR gives Death Nettle restrictions in exchange for being a powerful melee weapon and chemical injector.
Gloves are a fairly limited resource, generally only being found in Botanist's lockers, and they are not available in any vending machine.
If the player is not wearing these gloves specifically, they will take a minor amount of heat damage and caustic damage, with a small chance of being temporarily paralyzed.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
Expands the DamageOnInteract system and component to allow stunning (paralyzing) a user who takes damage when interacting with an entity with the DamageOnInteract component.
Also includes the option to add a delay on when a user can interact with an object with this component, primarily to stop spam clicking.
Also adds the ability to throw the object in a random direction if the entity is not being pulled. (Due to throwing not currently being predicted, this caused the entity to spastically jostle around)


https://github.com/user-attachments/assets/d5e1dd2c-d618-45aa-acb4-9ba26e5ea869



## Media
<!-- 
PRs which make in-game changes (adding clothing, items, new features, etc.) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

:cl:
- add: Botanist's leather gloves must now be equipped in order to pick up Death Nettles.
- tweak: Death Nettles will now pierce hardsuits, injecting 5 units of reagent.
- add: Death Nettles will now "wilt" after 5 hits.
- tweak: Fly Amanita has been changed to contain less Amatoxin.

Sirionaut commented:
5 damage seems a bit low if you don't use leather gloves, so low it heals on it own?(not sure where the threshold is).

I would say it should deal the same damage types as the nettles, so burn and caustic, and make it scale with potency. picking it up without gloves should be almost as bad as getting hit with it.

regarding the code, people will suggest to fix the whitespaces and probably take a look at this https://github.com/space-wizards/space-station-14/discussions/21113

Hmeister-real commented:
> 5 damage seems a bit low if you don't use leather gloves, so low it heals on it own?(not sure where the threshold is).
> 
> I would say it should deal the same damage types as the nettles, so burn and caustic, and make it scale with potency. picking it up without gloves should be almost as bad as getting hit with it.

I think it should inject 1/4 of what a hit would do + 5 burn,  since your not hitting yourself with it your just trying to hold it.

Kadeo64 commented:
botany should not be arming the masses with deathnettle, they should be arming the masses with gatfruits like the harvest gods intended

Arendian commented:
I still think the injection amount should be a bit lower than it originally was. If you swab amatoxin to it, it can crit someone in two hits(6u of amatoxin * 8 damage = 48 poison per hit) regardless of the armor they're wearing. 
With a few extremely rare weapons as an exception(throngler) there is not a single other melee weapon that comes even remotely close to this damage ratio. Even Energy swords need 4 hits to kill an unarmored target, more if the target is armored, and those cost 8 tc while deatnettles are basically free..

ViceEmargo commented:
> I still think the injection amount should be lower than it originally was. If you swab amatoxin to it, it can crit someone in two hits: 6u of amatoxin * 12 damage = 72 poison per hit regardless of the armor they're wearing + 7.5 heat and 7.5 caustic for a total of **89 damage!!** per hit on unarmored targets.
> 
> At an injection rate of 1.5 per hit instead of 6 it's still a very respectable 33 damage per hit on unarmored targets (1.5 * 12 = 18 poison from the amatoxin + 7.5 caustic + 7.5 heat = 33), which is 4 hits to crit someone. It will also make the deathnettle empty in 8 hits instead of 3 while keeping it's total damage potential the same but spread out over more hits, allowing you to crit two people(4 hits per person) with 1 deathnettle before it's empty.
> 
> This damage per hit is a tiny bit higher than the damage of an energy sword(15 heat + 15 slash), and I think deathnettles should at best be equal to one of those.
> 
> To summarize: with a few extremely rare weapons as an exception(throngler) there is not a single other melee weapon that comes even remotely close to the damage ratio of a deathnettle that injects 6u per hit. Even Energy swords need 4 hits to crit an unarmored target, more if the target is armored, and those cost 8 tc and are extremely loud, while deatnettles are basically free and pretty quiet. So I think the injection rate should be set to 1.5 instead of 6 to lower the damage per hit on unarmored targets to 33 instead of the current 89.

I think half or more of complaints about Death Nettle are really about Amatoxin.
I agree that putting amatoxin in death nettle in its current state is incredibly strong, but I don't believe changes to Death Nettle should be made simply due to what amatoxin is capable of making it.
That being said, after my experience making this PR, I would like to look into balancing Amatoxin itself.
I'd like to go into more detail when I look into it making a PR for it, but my thoughts right now are that there should be less Amatoxin in Amanita, (currently with 50 potency, you can get 13 amatoxin. I think closer to 6 or 8 would alleviate the power it can give to death nettle), and make Amatoxin (as well as other poisons to be honest) a slower metabolism rate.

UbaserB commented:
yeess i always wanted the requirement of wearing special gloves to use

PolterTzi commented:
I do think it would be fitting for both nettles to inject you with a bit of their reagents if handled improperly (without gloves). Maybe do let people use them without gloves, just at the cost of getting injected themselves each time they put it in their hands or attack.
But not sure if it's enough to justify reverting the nerfs.

ViceEmargo commented:
@Partmedia Is it okay that I mention you to review this PR since you seem to be involved in Botany with your PR?

Partmedia commented:
That's fine, yes, though I need to get up to speed on the death nettle discussion. Can you please link me to relevant discussions about this on Discord or from other PR's?

I skimmed your PR, and there are some code issues that we need to address, but we can worry about those once we conceptually agree how to go forward from here.

Anzarot121 commented:
> That's fine, yes, though I need to get up to speed on the death nettle discussion. Can you please link me to relevant discussions about this on Discord or from other PR's?
> 
> I skimmed your PR, and there are some code issues that we need to address, but we can worry about those once we conceptually agree how to go forward from here.

There are discussions here, here and here.
https://github.com/space-wizards/space-station-14/issues/25029
https://github.com/space-wizards/space-station-14/pull/25177
https://github.com/space-wizards/space-station-14/pull/25068
By the way, I must inform you that the PR of the second link essentially makes it so that at 100 potency, nettle contains only 3 histamines.
And also, author, add to the text of the PR that it solves the issue to which I provided a link.

ViceEmargo commented:
> That's fine, yes, though I need to get up to speed on the death nettle discussion. Can you please link me to relevant discussions about this on Discord or from other PR's?
> 
> I skimmed your PR, and there are some code issues that we need to address, but we can worry about those once we conceptually agree how to go forward from here.

Thank you very much. As well as some of the github discussions that Anzarot121 shared, here is a discussion thread on the Discord that is relevant. Although, this discord thread became quite bloated, and has a lot to read.

https://discord.com/channels/310555209753690112/1204095027836624956 - "Death nettles need some reeling in."

I am totally happy to accept any code criticism to improve the flow and adhere to accepted standards.

Partmedia commented on Content.Server/Botany/Components/ThornImmuneComponent.cs:
Why does this duplicate `ThornyImmuneComponent`?

Partmedia commented on Content.Server/Botany/Components/ThornyImmuneComponent.cs:
```diff
@@ -0,0 +1,7 @@
+public sealed partial class ThornyImmuneComponent : Component
+{
+    [DataField]
+    public bool ThornImmune = true;
```
`ThornyImmuneComponent` seems that it should automatically confer thron resistance, so this data field is not necessary.

Partmedia commented on Content.Server/Botany/Systems/ThornySystem.cs:
```diff
@@ -0,0 +1,56 @@
+        if (_inventory.TryGetSlotEntity(user, "gloves", out var slotEntity) &&
+            TryComp<ThornyImmuneComponent>(slotEntity, out var immunity))
+        {
+            hasImmunity = immunity.ThornImmune;
```
If you put the `return` here, you can get rid of the `hasImmunity` variable.

Partmedia commented on Content.Server/Botany/Systems/ThornySystem.cs:
```diff
@@ -0,0 +1,56 @@
+    [Dependency] private readonly DamageableSystem _damageableSystem = default!;
+    [Dependency] private readonly SharedPopupSystem _popupSystem = default!;
+    [Dependency] private readonly IAdminLogManager _adminLogger = default!;
+    public override void Initialize()
```
Add space above

Partmedia commented on Content.Server/Botany/Systems/ThornySystem.cs:
```diff
@@ -0,0 +1,56 @@
+        base.Initialize();
+        SubscribeLocalEvent<ThornyComponent, ContainerGettingInsertedAttemptEvent>(OnHandPickUp);
+    }
+    private void OnHandPickUp(Entity<ThornyComponent> entity, ref ContainerGettingInsertedAttemptEvent args)
```
Add space above

Erisfiregamer1 commented:
Accidental close or just gave up?

ViceEmargo commented:
I am closing this post until after feature freeze.
I saw PartMedia's code revision requests about 2 weeks after he posted them, then I made my code revisions, tested my code on my branch, and it was successful.
However, due to my lack of experience with github and using git commands, I was unable to get the changes pushed, and I grew a bit tired of trying to fix it.
I will sort it out and remake a PR after the feature freeze.

ViceEmargo commented:
I have made code revisions based on your review, and have removed an unnecessary key from the botanist's gloves in gloves.yml as pointed out by the YAML linter.

Nootaro commented:
Can't wait for people to lose their whole arm because they were not wearing gloves once newmed is done. :D

metalgearsloth reviewed (dismissed):
1. None of the code is documented.
2. None of it is predicted.
3. Needs localising.
4. Needs to be made generic and not hardcoded to one pickup (i.e. I'm pretty sure something else like glass does something similar).

ViceEmargo commented:
Notes to myself:
Document my code - Commenting guidelines https://docs.spacestation14.com/en/general-development/codebase-info/conventions.html

Learn how prediction is done (And what the difference between components/systems in Content.Server/Client are vs Content.Shared)

Learn what localization is (I think it means don't hard code human readable text, but reference it from a file?)

After peaking at Content.Shared/Damage/Components and Systems, I found DamageOnHoldingSystem / Component.
Perhaps utilize this and instead make death nettle deal damage over time? (I would still prefer the item be ejected from the hand though)

ViceEmargo commented:
It must have requested a review automatically, this is still very much a WIP as I figure out how to make move systems/components to be more generic, and figure out how to do prediction and localization.

slarticodefast commented:
> It must have requested a review automatically, this is still very much a WIP as I figure out how to make move systems/components to be more generic, and figure out how to do prediction and localization.

For localization, all you will have to do is use `Loc.GetString` for the popup instead of hardcoding the string. Here is a guide on how to do that:
https://docs.spacestation14.com/en/ss14-by-example/fluent-and-localization.html
You should also not hardcode the name of the nettle into the string, but automatically insert the entity name, so that if you add the same component to another entity, it will show up correctly.

For prediction you will have to move the component and corresponding system into shared, so both server and client can run it.

slarticodefast commented on Content.Server/Botany/Components/ThornyComponent.cs:
```diff
@@ -0,0 +1,16 @@
+    public string Sound = string.Empty;
+
+    [DataField]
+    public int ThrowStrength = 3;
```
Rename this one into ThrowSpeed, so its meaning is more clear. The same thing was recently done for the TryThrow function parameter.

slarticodefast commented on Content.Shared/Damage/Components/DamageOnPickupComponent.cs:
```diff
@@ -0,0 +1,10 @@
+public sealed partial class DamageOnPickupComponent : Component
+{
+
+}
```
It is probably fine if this is combined with the ThornyComponent, just make the throwing optional via a bool DataField

slarticodefast reviewed (commented):
Still WIP, but I did a partial review anyways with a few things I noticed :)

I'm looking forward to seeing this in-game once it is done.
Feel free to ask on Discord if you need help with anything.

slarticodefast commented:
Is this ready for re-review now?
Looks much simple now that it is using DamageOnInteract

ViceEmargo commented:
> Is this ready for re-review now? Looks much simple now that it is using DamageOnInteract

No, I think I should give survey the code again, and see if I can add more flavor to this, after I've changed to DamageOnInteract. I'm still concerned about adding throwing physics, as to the client, when they picked up the nettle and failed, the nettle would keep snapping to the player, then throwing.

slarticodefast commented:
Since throwing is not predicted the sudden "snapping" part might be unavoidable at the moment. An alternative idea you could try out instead of picking up and then throwing the item would be to cancel the pickup and throw the item from where it was lying. This might potentially be visually more appealing.

ViceEmargo commented:
> Since throwing is not predicted the sudden "snapping" part might be unavoidable at the moment. An alternative idea you could try out instead of picking up and then throwing the item would be to cancel the pickup and throw the item from where it was lying. This might potentially be visually more appealing.

Yeah, I did try that too, perhaps I'll try to re-add that.
I also JUST took a peak at SS13 botany to get some more ideas, and I just found out they actually have this exact feature I wanted of nettles and botanist's gloves. Apparently the nettle also had a chance to stun the user if they failed to pick it up.....and the nettle had a mechanic of wilting away with every use....I think I might look into implementing features like those if I can muster the gumption.

ViceEmargo commented:
Thank you for all the assistance and guidance.
This PR is ready for Re-Review

SlamBamActionman commented:
Hello! I've taken the time to go through this PR with how stuff works in master and such, and this is my feedback:

As has been observed, amatoxin is a lynchpin when it comes to balancing nettles. The non-penetration of the normal nettles has made it basically dead when it comes to antag work, despite the potential to deal 55 damage per nettle/swing. 

Making the death nettle able to penetrate armor is a suitable upgrade to make it distinct. However, that also makes it necessary to be much more careful when balancing it. As-is, I think this PR makes the death nettle waaaay too strong for pretty much anything, either for crew or antags. It's already very strong in its basic form, but combining that with amatoxin (even at only 50% content) is really scary (you can two-hit SecOffs with a 50% amatoxin yield, and *almost* two-hit a bloodred). 

I think making it comparable to an e-sword if properly upgraded is a suitable place to put it. For that, I propose making it into a 2u injector instead. This is less than the nettle but that is offset by the fact it's armor penetrating. 

This results in the following damages:
```
Base damage: 16 (8 heat, 8 caustic)
2u injection: 20 (16 + 2 + 2)
2u w/ 50% poison: 33 (16 + 15 + 2)
2u w/ 100% poison: 46 (16 + 30)

Sec Hardsuit:
Nothing inside: 13 (8 heat, 5 caustic)
2u injection: 17 (13 + 2 + 2)
2u w/ 50% poison: 30 (13 + 15 + 2)
2u w/ 100% poison: 43 (13 + 30)

Nukie:
Nothing inside: 8 (4 heat, 4 caustic)
2u injection: 12 (8 + 2 + 2)
2u w/ 50% poison: 25 (8 + 15 + 2)
2u w/ 100% poison: 38 (8 + 30)
```

This means that a botanist who is going through the effort of growing a death nettle and perfecting it to a near 100% amatoxin potency will come out on top against an e-sword, saving 8 TC and gaining other benefits such as being far less obvious. It will still be a good weapon against nukies if the poison is somewhat refined, but it's offset by being a melee weapon and difficulty to access. 

And if you don't refine it? A 20 per hit weapon against unarmored targets for just a bit of mutating is a perfectly fine result.

ViceEmargo commented:
> Hello! I've taken the time to go through this PR with how stuff works in master and such, and this is my feedback:
> 
> As has been observed, amatoxin is a lynchpin when it comes to balancing nettles. The non-penetration of the normal nettles has made it basically dead when it comes to antag work, despite the potential to deal 55 damage per nettle/swing.
> 
> Making the death nettle able to penetrate armor is a suitable upgrade to make it distinct. However, that also makes it necessary to be much more careful when balancing it. As-is, I think this PR makes the death nettle waaaay too strong for pretty much anything, either for crew or antags. It's already very strong in its basic form, but combining that with amatoxin (even at only 50% content) is really scary (you can two-hit SecOffs with a 50% amatoxin yield, and _almost_ two-hit a bloodred).
> 
> I think making it comparable to an e-sword if properly upgraded is a suitable place to put it. For that, I propose making it into a 2u injector instead. This is less than the nettle but that is offset by the fact it's armor penetrating.
> 
> This results in the following damages:
> 
> ```
> Base damage: 16 (8 heat, 8 caustic)
> 2u injection: 20 (16 + 2 + 2)
> 2u w/ 50% poison: 33 (16 + 15 + 2)
> 2u w/ 100% poison: 46 (16 + 30)
> 
> Sec Hardsuit:
> Nothing inside: 13 (8 heat, 5 caustic)
> 2u injection: 17 (13 + 2 + 2)
> 2u w/ 50% poison: 30 (13 + 15 + 2)
> 2u w/ 100% poison: 43 (13 + 30)
> 
> Nukie:
> Nothing inside: 8 (4 heat, 4 caustic)
> 2u injection: 12 (8 + 2 + 2)
> 2u w/ 50% poison: 25 (8 + 15 + 2)
> 2u w/ 100% poison: 38 (8 + 30)
> ```
> 
> This means that a botanist who is going through the effort of growing a death nettle and perfecting it to a near 100% amatoxin potency will come out on top against an e-sword, saving 8 TC and gaining other benefits such as being far less obvious. It will still be a good weapon against nukies if the poison is somewhat refined, but it's offset by being a melee weapon and difficulty to access.
> 
> And if you don't refine it? A 20 per hit weapon against unarmored targets for just a bit of mutating is a perfectly fine result.

Hello,
Thank you very much for this information.
I am aware of the unholy alliance between Death Nettles and amanita shrooms.
Perhaps I should have included this relevant information when creating this PR long ago.
Amatoxin used to deal slightly less damage per unit at the time of creating this PR, although faster.

I would like to balance Death Nettle itself based on its default circumstances, as it includes Sulfuric Acid and Fluorosulfuric acid.
Death Nettle in this PR only gets 5 swings before it it destroyed. All 5 swings will just put a person in a Security Hardsuit into crit, do about 80 damage to a person in a Blood-red hardsuit, and crit a person not wearing any armor in 4 hits.

I recognize that amatoxin is wildly superior to the acids in damage to unit,

1u of Fluorosulfuric acid is 2.6 Caustic damage basically immediately
1u of Surfuric Acid is 1.6 Caustic damage basically immediately
A 50 potency Death Nettle will contain 9.33 of sulfuric acid, and 9.33 of fluorosurfuric acid.
The sum total damage of these injected reagents is roughly 40 Caustic damage.

A 50 potency Fly Amanita Spore contains 13 amatoxin.
If you swab this into Death Nettle, this 13 amatoxin will give you a damage potential of 195 Poison damage (holy smokes, eh?)

Regardless if the Death Nettle injects 2u or 5u, this damage potential is scary.
Death Nettle is not a loud weapon, and is frankly a silent killer (ignore the screams of your enemies as their blood is eroded by highly potent acids!)

I was hoping to work on this PR first, but Fly Amanita Spores are my primary target for change next.
A 50 potency Fly Amanita Spore can round remove anyone who happens to snack on one. (Given the circumstances that someone eats a shroom, and when they enter crit, they are injected with an airloss medication)
I'm hoping to both reduce the metabolism rate for amatoxin, lower the damage per tick, and severely reduce the amount of amatoxin contained within a single shroom.
The power of amatoxin swabbed into Death Nettle lies in that 50 Potency. 13 amatoxin is a lot.
I'd like to change that down to somewhere between 3 and 5 at 50 potency.
In amatoxin's current state, if a 50 potency fly amanita spore were swabbed into Death Nettle, that would only give a potential damage of 75 Poison damage, still a lot, but much more reasonable.

TL;DR, 
I don't want to punish Death Nettle for the great sin of Fly Amanita Spores.
Instead, I would like to tackle Fly Amanita Spores in a separate, but very necessary PR.
As this was my first PR, I was hesitant to open any other PRs, as this is very much a learning experience on the workflow.

SlamBamActionman commented:
Since Death Nettle and Amatoxin are so heavily intwined I brought it up in the maintainer discussion;

Independent of the Death Nettle we feel the Amatoxin itself is in a pretty good place; it's already slower than other toxins so it's not a great stealth kill poison, so with its potency it fills a niche as a great WarOps spear enhancer or to send someone to medbay to be looted/further injected.

The fact that the Amanita spores can have such high reagent amount is where we found issues, and where we think a nerf is appropriate. Were the death nettles to be nerfed to 2u injection & 5 hits it is unreasonable that a basic 50% potency amanita fills that entire use (13.5u). 

After discussing it we feel a nerf to the reagent amount is warranted, since that can affect how it's used in nettles but doesn't impact its use elsewhere much. The suggested amount was a max of 10u amatoxin at 100% potency; that'll leave 5u at 50%, so while it is still a strong combo at that potency it is no longer as sustainable. It'll let experienced botanists try to tinker for a perfect 100% potency nettle which will give all five hits full impact, which rewards the time investment and risk of getting caught (since both death nettles and amanita spores are highly suspicious to grow).

Considering the relationship nettles and amatoxin has, I think it's preferable if the nerf is included in this PR. It's just a single value change of the max reagent amount from 25 to 10 after all.

ViceEmargo commented:
Honestly, it is nigh impossible to reach 100 potency. It is already difficult just to get 50 potency without destroying a plant because the kudzu and death gene they can contract are so easy to roll. 50 potency is where stuff like this showed primarily be balanced, as it is easily obtainable.
I've adjusted the values (seeds.yml) so that Fly Amanita Spores will contain 5.16 amatoxin (and 3.5 nutriment) at 50 potency, and 9.33 amatoxin at 100 potency.
I've adjusted the entity itself (spawned from the entity menu) as well.

UbaserB commented:
> kudzu and death gene they can contract are so easy to roll.

Just wanted to point out, with enough chems basically anything is possible.. since there's phalanx which removes the death gene, sedin which removes the seedless gene, alongside healing chems like ammonia/dylo/etc. Sure, its relatively hard to get but I feel like getting to a high amount of potency isn't actually all *that* difficult.

ViceEmargo commented:
Maybe I'm just a cynical botanist, but it can be hard to acquire those chemicals from the Chemistry staff.
I've personally never even seen Sedin (although, I have never asked for it.)
The highest potency I've gotten is about 58, and the highest I've seen from someone else was somewhere in the mid 70s.

Also, updated my branch to latest upstream seemed to fix the build issue.

ViceEmargo commented:
Just noticed a Debug.Log statement was not removed. I will remove this and push another commit soon.

Pumkin69 commented:
goated pr

SlamBamActionman reviewed (approved):
I expect us to see parts of this PR be nerfed once Botany finds out some way to turn Death Nettles into the Throngler, but for now, let's see what happens! Thank you for your contribution!

slarticodefast commented on Content.Shared/Damage/Components/DamageOnInteractComponent.cs:
```diff
@@ -44,4 +44,46 @@ public sealed partial class DamageOnInteractComponent : Component
+    /// Time between being able to interact with this entity
+    /// </summary>
+    [DataField]
+    public uint InteractTimer = 0;
```
should be a `TimeSpan`

slarticodefast commented on Content.Shared/Damage/Components/DamageOnInteractComponent.cs:
```diff
@@ -44,4 +44,46 @@ public sealed partial class DamageOnInteractComponent : Component
+    /// <summary>
+    /// Tracks the last time this entity was interacted with, but only if the interaction resulted in the user taking damage
+    /// </summary>
+    [DataField]
+    public TimeSpan LastInteraction = TimeSpan.Zero;
+
+    /// <summary>
+    /// Tracks the time that this entity can be interacted with, but only if the interaction resulted in the user taking damage
+    /// </summary>
+    [DataField]
+    public TimeSpan NextInteraction = TimeSpan.Zero;
```
You only need one of these two, the other one can be calculated

slarticodefast commented on Content.Shared/Damage/Systems/DamageOnInteractSystem.cs:
```diff
@@ -35,6 +45,13 @@ public override void Initialize()
+        // Stop the interaction if the user attempts to interact with the object before the timer is finished
+        if (_gameTiming.CurTime < entity.Comp.NextInteraction)
+        {
+            args.Handled = true;
```
This will stop all other interactions subscribing to `InteractHandEvent`.
But only those that are done after this subscription, which is bad because the order in which they are handled is random.

beck-thompson replied:
I was going to review this, but the event doesn't handle all the situations correctly. Like if you pick up the item, and take the gloves of it is fine with that. There was one other case similar as well but I forget :thinking:

slarticodefast commented on Content.Shared/Damage/Systems/DamageOnInteractSystem.cs:
```diff
@@ -59,17 +75,31 @@ private void OnHandInteract(Entity<DamageOnInteractComponent> entity, ref Intera
         if (totalDamage != null && totalDamage.AnyPositive())
         {
+            // Record this interaction and determine when a user is allowed to interact with this entity again
+            entity.Comp.LastInteraction = _gameTiming.CurTime;
+            entity.Comp.NextInteraction = _gameTiming.CurTime + TimeSpan.FromSeconds(entity.Comp.InteractTimer);
```
You only need one of these.
Also this eats the remainder every time it is updated.

slarticodefast commented on Content.Shared/Damage/Systems/DamageOnInteractSystem.cs:
```diff
@@ -59,17 +75,31 @@ private void OnHandInteract(Entity<DamageOnInteractComponent> entity, ref Intera
         }
+        // Check if the entity's Throw bool is false, or if the entity has the PullableComponent, then if the entity is currently being pulled.
+        // BeingPulled must be checked because the entity will be spastically thrown around without this.
+        if (!entity.Comp.Throw || !TryComp<PullableComponent>(entity, out var pullComp) || pullComp.BeingPulled)
```
shouldn't this be part of TryThrow in general?

slarticodefast commented on Content.Shared/Damage/Systems/DamageOnInteractSystem.cs:
```diff
@@ -59,17 +75,31 @@ private void OnHandInteract(Entity<DamageOnInteractComponent> entity, ref Intera
+        if (!entity.Comp.Throw || !TryComp<PullableComponent>(entity, out var pullComp) || pullComp.BeingPulled)
+            return;
+
+        _throwingSystem.TryThrow(entity, _random.NextVector2(), entity.Comp.ThrowSpeed, doSpin: true);
```
this will genare a different random vector on the client and the server
